File: src/graph_mcp/rate_limiter.py
from __future__ import annotations

import asyncio
import time
from collections import deque

from graph_mcp.exceptions import RateLimitError


class RateLimiter:
    def __init__(self, max_requests: int = 10000, window: int = 600):
        self.max_requests = max_requests
        self.window = window
        self._timestamps: deque[float] = deque()
        self._backoff_until: float = 0.0
        self._backoff_count: int = 0

    def _clean_window(self) -> None:
        cutoff = time.monotonic() - self.window
        while self._timestamps and self._timestamps[0] < cutoff:
            self._timestamps.popleft()

    async def acquire(self) -> None:
        now = time.monotonic()
        if now < self._backoff_until:
            wait = self._backoff_until - now
            await asyncio.sleep(wait)

        self._clean_window()
        if len(self._timestamps) >= self.max_requests:
            raise RateLimitError(
                f"Rate limit exceeded: {self.max_requests} requests "
                f"per {self.window}s window"
            )
        self._timestamps.append(time.monotonic())

    def handle_429(self, retry_after: float | None = None) -> float:
        self._backoff_count += 1
        delay = retry_after or min(2**self._backoff_count, 60)
        self._backoff_until = time.monotonic() + delay
        return delay

    def reset_backoff(self) -> None:
        self._backoff_count = 0
        self._backoff_until = 0.0


_rate_limiter: RateLimiter | None = None


def get_rate_limiter() -> RateLimiter:
    global _rate_limiter
    if _rate_limiter is None:
        from graph_mcp.config import settings

        _rate_limiter = RateLimiter(
            max_requests=settings.graph_rate_limit_max_requests,
            window=settings.graph_rate_limit_window,
        )
    return _rate_limiter

File: src/rate-limiter.ts
import { RateLimitError } from "./errors.js";

export interface RateLimiterOptions {
  maxRequests: number;
  windowMs: number;
  now?: () => number;
  sleep?: (milliseconds: number) => Promise<void>;
}

const defaultNow = (): number => Number(process.hrtime.bigint()) / 1_000_000;

const defaultSleep = (milliseconds: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, milliseconds);
  });

const MAX_TIMER_DELAY_MS = 2_147_483_647;

export class RateLimiter {
  private readonly maxRequests: number;
  private readonly windowMs: number;
  private readonly now: () => number;
  private readonly sleep: (milliseconds: number) => Promise<void>;
  private readonly timestamps: number[] = [];
  private backoffUntil = 0;
  private backoffCount = 0;

  constructor(options: RateLimiterOptions) {
    this.maxRequests = options.maxRequests;
    this.windowMs = options.windowMs;
    this.now = options.now ?? defaultNow;
    this.sleep = options.sleep ?? defaultSleep;
  }

  async acquire(): Promise<void> {
    let now = this.now();
    while (now < this.backoffUntil) {
      await this.sleep(this.backoffUntil - now);
      now = this.now();
    }

    const cutoff = now - this.windowMs;
    while (this.timestamps.length > 0) {
      const timestamp = this.timestamps[0];
      if (timestamp === undefined || timestamp >= cutoff) {
        break;
      }
      this.timestamps.shift();
    }

    if (this.timestamps.length >= this.maxRequests) {
      throw new RateLimitError(
        `Rate limit exceeded: ${this.maxRequests} requests per ${this.windowMs / 1000}s window`,
      );
    }

    this.timestamps.push(this.now());
  }

  handle429(retryAfterSeconds?: number): number {
    this.backoffCount += 1;
    const retryAfterMilliseconds =
      retryAfterSeconds !== undefined && Number.isFinite(retryAfterSeconds) && retryAfterSeconds > 0
        ? Math.min(retryAfterSeconds * 1000, MAX_TIMER_DELAY_MS)
        : undefined;
    const delayMilliseconds = retryAfterMilliseconds ?? Math.min(2 ** this.backoffCount, 60) * 1000;
    const now = this.now();
    const proposedDeadline = now + delayMilliseconds;
    this.backoffUntil = Math.max(this.backoffUntil, proposedDeadline);
    return (this.backoffUntil - now) / 1000;
  }

  resetBackoff(): void {
    this.backoffCount = 0;
    this.backoffUntil = 0;
  }
}
